### PR TITLE
Add user access profiles

### PR DESCRIPTION
**Pull request by lgarvey**

### CODE REVIEW

File: sso/user/models.py [hunk 1]
```diff
@@ -128,6 +163,22 @@ def can_access(self, application):
         else:
             return application in self.permitted_applications.all()
 
+    def can_access(self, application: OAuth2Application):
+        """ Can the user access this application?"""
+
+        # check if an AccessProfile gives the user access
+        for profile in self.access_profiles.all():
```
Comment (michalc) [inline sso/user/models.py]: So this gives _really_ predictable behaviour, maybe this should be sorted somehow? So _exactly_ the same thing happens for a given users's config every time, and it's not up to the database on the order it returns the access_profiles?

File: sso/tests/test_user.py [hunk 2]
```diff
@@ -229,6 +229,17 @@ def test_can_access_without_perms(self):
...
+    @pytest.mark.django_db
+    def test_can_access_with_access_profile(self):
+
+        app = ApplicationFactory()
+        user = UserFactory()
+        ap = AccessProfile.objects.create(name='test profile')
+        ap.oauth2_applications.add(app)
+        user.access_profiles.add(ap)
+
+        assert user.can_access(app)
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply sso/tests/test_user.py]: Yep, made senes and I've done this.

File: sso/user/models.py [hunk 3]
```diff
@@ -128,6 +163,22 @@ def can_access(self, application):
...
+            if profile.is_allowed(application):
+                return True
+
+        # check if the user is directly granted access via self.permitted_applications or
+        # application configuration: default_access_allowed or allow_access_by_email_suffix
+        if self._is_user_permitted_to_access_application(application):
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply sso/user/models.py]: Yep, agreed also - it wasn't so well thought out.  I've refactored it into a single method.

File: sso/user/models.py (hunk 1 above)
Comment (lgarvey) [reply sso/user/models.py]: I don't think this is so important. It may be if/when the logic changes from "permitted by any profile"

**Pull request merged**